### ISSUE TITLE
allowed chunksize seems much smaller than worker memory limit

### ISSUE THREAD

**Issue by phildong** [dask/dask#6677]
**What happened**:
This may be a bug but could also be just me not understanding dask properly, but here's what happened:

I'm creating a random dask array and just directly saving it to zarr, and I'm using a LocalCluster with certain memory limit. As far as I can understand, dask should not have a issue with memory as long as the memory demand of computation on each chunk is smaller than the worker memory limit. And in this simple case it should be fine as long as the chunksize is, say smaller than 90% of the worker memory limit, since all it's happening is generating some random numbers and saving it directly to disk. However that doesn't seem to be true. The "threshold" of working chunk size seem to be around 1/8 ~ 1/10 of the worker memory limit. Whenever I set chunk size larger than that threshold, I start to get the "Worker paused/killed due to memory limit" and "memory limit reached but cannot save cache to disk, maybe there's a memory leak?" warning/messages.

**What you expected to happen**:
The tasks should finish as long as chunk size is smaller than worker memory limit.

**Minimal Complete Verifiable Example**:

```python
from dask.distributed import Client
import dask.array as darr

client = Client(memory_limit='1GB')

a1 = darr.random.random(1e9, chunks=100*1000**2/8) # 100MB chunk
a2 = darr.random.random(1e9, chunks=125*1000**2/8) # 125MB chunk
a3 = darr.random.random(1e9, chunks=150*1000**2/8) # 150MB chunk

a1.to_zarr("test1.zarr", overwrite=True) # this will finish properly
a2.to_zarr("test2.zarr", overwrite=True) # this will throw "worker exceeds 95% limit" warnings but still may finish
a3.to_zarr("test3.zarr", overwrite=True) # this will always lead to "worker killed" and eventually exceptions
```

**Anything else we need to know?**:
1. I tried doubling the memory limit, and "max working chunksize" seem to scale proportionally, instead of a fixed amount (~900MB) smaller than the memory limit.
1. I also tried disabling work stealing but that doesn't seem to help.

**Environment**:

- Dask version: 2.28.0
- Python version: 3.8.5
- Operating System: Ubuntu 18.04.5
- Install method (conda, pip, source): conda-forge

**Comment by mrocklin** [comment]
This might help:
https://docs.dask.org/en/latest/best-practices.html#avoid-very-large-partitions

**Comment by phildong** [comment]
Ha got it! missed that part of the documentation.
Some follow up questions:
1. So it seems 1/10 of worker memory is a generally good chunksize. Is that correct?
1. From the documentation it seems dask stores more than one chunk per worker, but does it try to carry out computation as well for more than one chunk at the same time? For example, if my function on each chunk creates a copy of the chunk as intermediate variable during the computation, should I account for it by using a chunksize of 1/20, or 1/11 of the worker memory?
1. Is there any possibility to disable this feature so that each worker strictly only deal with one chunk at a time? Could that be helpful in memory-bound tasks?

**Comment by mrocklin** [comment]
1.  It depends on the number of concurrent threads running
2.  yes
3.  You could use workers with only one thread, even then though there
might be an intermediate chunk lying around.

**Comment by phildong** [comment]
Ok sorry if I'm missing the point, but I'm still confused about how to find the best chunksize that guarantee success. Specifically:

From your comments I gathered there are two things governing "additional" memory demands: 1. number of threads per worker, which determines the number of computations that will run concurrently, and can be controlled with `nthreads` argument. 2. number of additional chunks a worker decide to keep around.

So really the total memory demand of a worker is `mem_demand_per_chunk * nthreads + chunksize * X_additional_chunk`?

If the above is correct I have 2 more questions:
1. I tried setting `nthreads=1` but it does not increase the chunksize I can use in my example. Does this mean the second part of the memory demand is much more significant in this case?
2. How do I find `X_additional_chunk` to make sure my tasks can finish without an error?

**Comment by mrocklin** [comment]
I recommend looking at the dashboard during computation to see if you can
build intuition about what is happening.  There is a video here:
https://docs.dask.org/en/latest/diagnostics-distributed.html

**Comment by DeniseCaiLab** [comment]
Yes I have been watching the dashboard for a while, all I can tell is whenever I set the chunksize beyond the "safe threshold" I find, workers start dying and new ones are created, so the task streams look funky. And regardless of the chunksize, the memory monitor of each worker always goes between blue and orange, so it seems they're trying to clear the cache, just not successfully if the chunksize is too large.

Am I missing anything? How would this help me find out what determines the max "safe" chunksize and whether there's something I can do to change the max "safe" chunksize?

**Comment by phildong** [comment]
The following failed too, it seems even dask itself isn't able to figure out a good chunk size??
```python
from dask.distributed import Client
client = Client(memory_limit='500MB', nthreads=1)
a = darr.random.random(1e9, chunks='auto')
a.to_zarr('test.zarr', overwrite=True)
```
I'm really lost on what we are supposed to do even in such simple cases? Is the official advice on this just to set the chunk size to 1/10 of your worker memory limit and hope it'll work? and if not keep decreasing the chunksize until it works?

**Comment by mrocklin** [comment]
I ran the following and it seemed to stay within about 3GB of memory.

500MB is quite small.  Just importing numpy, pandas, python, zarr, and so on typically takes up a few hundred MB in each process,

**Comment by phildong** [comment]
I believe the main issue here is dask is throwing in too much tasks in memory according to the number of threads you have. Once I properly limit the number of threads like the following I can get it to run within 1GB of memory, which at least falls in the "10x chunk size" range.
```python
import dask.array as darr
from dask.distributed import LocalCluster, Client
cluster = LocalCluster(
    n_workers=1,
    memory_limit="1GB",
    threads_per_worker=1,
)
client = Client(cluster)
a = darr.random.random(1e9, chunks='auto')
a.to_zarr('test.zarr', overwrite=True)
```
Closing this now as I consider limiting threads as a workaround for this particular issue. I posted a sequel at dask/distributed#4491 regarding how to have more control over the number of concurrent jobs.
